spring-restdocs/src/main/java/org/springframework/restdocs/hypermedia/FieldSnippetResultHandler.java

```diff
@@ -0,0 +1,125 @@
+import org.springframework.util.Assert;
+
+/**
+ * A {@link SnippetWritingResultHandler} that produces a snippet documenting a RESTful
+ * resource's request fields.
+ * 
+ * @author Andreas Evers
+ */
+public class FieldSnippetResultHandler extends SnippetWritingResultHandler {
+
+	private final Map<String, FieldDescriptor> descriptorsByName = new HashMap<String, FieldDescriptor>();
+	
+	private final FieldExtractor extractor;
+	private final Type type;
+	
+	enum Type {
+		REQUEST, RESPONSE;
+	}
+
+	FieldSnippetResultHandler(String outputDir, FieldExtractor fieldExtractor,
+			FieldSnippetResultHandler.Type type, List<FieldDescriptor> descriptors) {
+		super(outputDir, type.toString().toLowerCase() + "fields");
+		this.extractor = fieldExtractor;
+		this.type = type;
+		for (FieldDescriptor descriptor : descriptors) {
+			Assert.hasText(descriptor.getName());
+			Assert.hasText(descriptor.getDescription());
+			this.descriptorsByName.put(descriptor.getName(), descriptor);
+		}
+	}
+
+	@Override
+	protected void handle(MvcResult result, DocumentationWriter writer)
+			throws IOException {
+		Map<String, Field> fields;
+		if (this.extractor != null) {
+			fields = this.extractor.extractFields(result.getResponse());
+		}
+		else {
+			String contentType = result.getResponse().getContentType();
+			FieldExtractor extractorForContentType = FieldExtractors
+					.extractorForContentType(contentType);
+			if (extractorForContentType != null) {
+				fields = extractorForContentType.extractFields(result.getResponse());
+			}
+			else {
+				throw new IllegalStateException(
+						"No FieldExtractor has been provided and one is not available for the content type "
+								+ contentType);
+			}
+
+		}
+		
+		Set<String> actualFields = fields.keySet();
+		Set<String> expectedFields = this.descriptorsByName.keySet();
+
+		Set<String> undocumentedFields = new HashSet<String>(actualFields);
+		undocumentedFields.removeAll(expectedFields);
+
+		Set<String> missingFields = new HashSet<String>(expectedFields);
+		missingFields.removeAll(actualFields);
+
+		if (!undocumentedFields.isEmpty() || !missingFields.isEmpty()) {
+			String message = "";
+			if (!undocumentedFields.isEmpty()) {
+				message += "Fields with the following names were not documented: "
+						+ undocumentedFields;
+			}
+			if (!missingFields.isEmpty()) {
+				message += "Fields with the following names were not found in the " + 
+						this.type.toString().toLowerCase() + ": "
+						+ missingFields;
+			}
+			fail(message);
+		}
+
+		Assert.isTrue(actualFields.equals(expectedFields));
+
+		writer.println("|===");
+		writer.println("| Name | Description | Type | Required | Constraints");
+
+		for (Entry<String, FieldDescriptor> entry : this.descriptorsByName.entrySet()) {
+			writer.println();
+			writer.println("| " + entry.getKey());
+			writer.println("| " + entry.getValue().getDescription());
+			writer.println("| " + entry.getValue().getType().getSimpleName());
```

Comment:
The documentation shouldn't include the name of a Java type – the fact that the service is written in Java is an implementation detail. Assuming a JSON-based format (which is entirely reasonable) the type can be one of `number`, `string`, `boolean`, `array`, or `object`.